File: formal/coqimp/task05/extraction_hs/Floats.hs
module Floats where

import qualified Prelude
import qualified Archi
import qualified BinPos
import qualified Coqlib
import qualified Datatypes
import qualified Fappli_IEEE
import qualified Fappli_IEEE_bits
import qualified Fappli_IEEE_extra
import qualified Fcore_Zaux
import qualified Integers
import qualified Specif

type Coq_float = Fappli_IEEE_bits.Coq_binary64

cmp_of_comparison :: Integers.Coq_comparison -> (Prelude.Maybe Datatypes.Coq_comparison) ->
                     Prelude.Bool
cmp_of_comparison c x =
  case c of {
   Integers.Ceq ->
    case x of {
     Prelude.Just c0 -> case c0 of {
                         Datatypes.Eq -> Prelude.True;
                         _ -> Prelude.False};
     Prelude.Nothing -> Prelude.False};
   Integers.Cne ->
    case x of {
     Prelude.Just c0 -> case c0 of {
                         Datatypes.Eq -> Prelude.False;
                         _ -> Prelude.True};
     Prelude.Nothing -> Prelude.True};
   Integers.Clt ->
    case x of {
     Prelude.Just c0 -> case c0 of {
                         Datatypes.Lt -> Prelude.True;
                         _ -> Prelude.False};
     Prelude.Nothing -> Prelude.False};
   Integers.Cle ->
    case x of {
     Prelude.Just c0 -> case c0 of {
                         Datatypes.Gt -> Prelude.False;
                         _ -> Prelude.True};
     Prelude.Nothing -> Prelude.False};
   Integers.Cgt ->
    case x of {
     Prelude.Just c0 -> case c0 of {
                         Datatypes.Gt -> Prelude.True;
                         _ -> Prelude.False};
     Prelude.Nothing -> Prelude.False};
   Integers.Cge ->
    case x of {
     Prelude.Just c0 -> case c0 of {
                         Datatypes.Lt -> Prelude.False;
                         _ -> Prelude.True};
     Prelude.Nothing -> Prelude.False}}

_Float__transform_quiet_pl :: Fappli_IEEE.Coq_nan_pl -> Fappli_IEEE.Coq_nan_pl
_Float__transform_quiet_pl pl =
  BinPos._Pos__lor (Specif.proj1_sig pl)
    (Fcore_Zaux.iter_nat (\x -> (\x -> 2 Prelude.* x) x) (Prelude.succ (Prelude.succ (Prelude.succ
      (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ
      (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ
      (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ
      (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ
      (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ
      (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ
      (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ
      (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ (Prelude.succ
      0))))))))))))))))))))))))))))))))))))))))))))))))))) 1)

_Float__neg_pl :: Prelude.Bool -> Fappli_IEEE.Coq_nan_pl -> (,) Prelude.Bool Fappli_IEEE.Coq_nan_pl
_Float__neg_pl s pl =
  (,) (Prelude.not s) pl

_Float__abs_pl :: Prelude.Bool -> Fappli_IEEE.Coq_nan_pl -> (,) Prelude.Bool Fappli_IEEE.Coq_nan_pl
_Float__abs_pl _ pl =
  (,) Prelude.False pl

_Float__binop_pl :: Fappli_IEEE_bits.Coq_binary64 -> Fappli_IEEE_bits.Coq_binary64 -> (,)
                    Prelude.Bool Fappli_IEEE.Coq_nan_pl
_Float__binop_pl x y =
  case x of {
   Fappli_IEEE.B754_nan s1 pl1 ->
    case y of {
     Fappli_IEEE.B754_nan s2 pl2 ->
      case Archi.choose_binop_pl_64 s1 pl1 s2 pl2 of {
       Prelude.True -> (,) s2 (_Float__transform_quiet_pl pl2);
       Prelude.False -> (,) s1 (_Float__transform_quiet_pl pl1)};
     _ -> (,) s1 (_Float__transform_quiet_pl pl1)};
   _ ->
    case y of {
     Fappli_IEEE.B754_nan s2 pl2 -> (,) s2 (_Float__transform_quiet_pl pl2);
     _ -> Archi.default_pl_64}}

_Float__zero :: Coq_float
_Float__zero =
  Fappli_IEEE.B754_zero Prelude.False

_Float__neg :: Coq_float -> Coq_float
_Float__neg =
  Fappli_IEEE.coq_Bopp ((\x -> x) ((\x -> 2 Prelude.* x Prelude.+ 1) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x Prelude.+ 1) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x Prelude.+ 1)
    1)))))) ((\x -> x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) 1))))))))))) _Float__neg_pl

_Float__abs :: Coq_float -> Coq_float
_Float__abs =
  Fappli_IEEE.coq_Babs ((\x -> x) ((\x -> 2 Prelude.* x Prelude.+ 1) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x Prelude.+ 1) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x Prelude.+ 1)
    1)))))) ((\x -> x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) 1))))))))))) _Float__abs_pl

_Float__add :: Coq_float -> Coq_float -> Coq_float
_Float__add =
  Fappli_IEEE.coq_Bplus ((\x -> x) ((\x -> 2 Prelude.* x Prelude.+ 1) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x Prelude.+ 1) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x Prelude.+ 1)
    1)))))) ((\x -> x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) 1)))))))))))
    _Float__binop_pl Fappli_IEEE.Coq_mode_NE

_Float__sub :: Coq_float -> Coq_float -> Coq_float
_Float__sub =
  Fappli_IEEE.coq_Bminus ((\x -> x) ((\x -> 2 Prelude.* x Prelude.+ 1) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x Prelude.+ 1) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x Prelude.+ 1)
    1)))))) ((\x -> x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) 1)))))))))))
    _Float__binop_pl Fappli_IEEE.Coq_mode_NE

_Float__mul :: Coq_float -> Coq_float -> Coq_float
_Float__mul =
  Fappli_IEEE.coq_Bmult ((\x -> x) ((\x -> 2 Prelude.* x Prelude.+ 1) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x Prelude.+ 1) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x Prelude.+ 1)
    1)))))) ((\x -> x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) 1)))))))))))
    _Float__binop_pl Fappli_IEEE.Coq_mode_NE

_Float__div :: Coq_float -> Coq_float -> Coq_float
_Float__div =
  Fappli_IEEE.coq_Bdiv ((\x -> x) ((\x -> 2 Prelude.* x Prelude.+ 1) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x Prelude.+ 1) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x Prelude.+ 1)
    1)))))) ((\x -> x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) 1)))))))))))
    _Float__binop_pl Fappli_IEEE.Coq_mode_NE

_Float__compare :: Coq_float -> Coq_float -> Prelude.Maybe Datatypes.Coq_comparison
_Float__compare f1 f2 =
  Fappli_IEEE.coq_Bcompare ((\x -> x) ((\x -> 2 Prelude.* x Prelude.+ 1) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x Prelude.+ 1) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x Prelude.+ 1)
    1)))))) ((\x -> x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) 1))))))))))) f1 f2

_Float__cmp :: Integers.Coq_comparison -> Coq_float -> Coq_float -> Prelude.Bool
_Float__cmp c f1 f2 =
  cmp_of_comparison c (_Float__compare f1 f2)

_Float__to_long :: Coq_float -> Prelude.Maybe Integers.Int64__Coq_int
_Float__to_long f =
  Coqlib.option_map Integers._Int64__repr
    (Fappli_IEEE_extra.coq_ZofB_range ((\x -> x) ((\x -> 2 Prelude.* x Prelude.+ 1)
      ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x Prelude.+ 1) ((\x -> 2 Prelude.* x)
      ((\x -> 2 Prelude.* x Prelude.+ 1) 1)))))) ((\x -> x) ((\x -> 2 Prelude.* x)
      ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
      ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
      ((\x -> 2 Prelude.* x) 1))))))))))) f Integers._Int64__min_signed Integers._Int64__max_signed)

_Float__from_parsed :: Prelude.Integer -> Prelude.Integer -> Prelude.Integer -> Coq_float
_Float__from_parsed base intPart expPart =
  Fappli_IEEE_extra.coq_Bparse ((\x -> x) ((\x -> 2 Prelude.* x Prelude.+ 1) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x Prelude.+ 1) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x Prelude.+ 1)
    1)))))) ((\x -> x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x)
    ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) ((\x -> 2 Prelude.* x) 1))))))))))) base intPart
    expPart

File: formal/coqimp/task01/extraction_ml/Floats.ml
open Archi
open BinNums
open BinPos
open Coqlib
open Datatypes
open Fappli_IEEE
open Fappli_IEEE_bits
open Fappli_IEEE_extra
open Fcore_Zaux
open Integers

type float = binary64

(** val cmp_of_comparison : comparison -> Datatypes.comparison option -> bool **)

let cmp_of_comparison c x =
  match c with
  | Ceq -> (match x with
            | Some c0 -> (match c0 with
                          | Eq -> true
                          | _ -> false)
            | None -> false)
  | Cne -> (match x with
            | Some c0 -> (match c0 with
                          | Eq -> false
                          | _ -> true)
            | None -> true)
  | Clt -> (match x with
            | Some c0 -> (match c0 with
                          | Lt -> true
                          | _ -> false)
            | None -> false)
  | Cle -> (match x with
            | Some c0 -> (match c0 with
                          | Gt -> false
                          | _ -> true)
            | None -> false)
  | Cgt -> (match x with
            | Some c0 -> (match c0 with
                          | Gt -> true
                          | _ -> false)
            | None -> false)
  | Cge -> (match x with
            | Some c0 -> (match c0 with
                          | Lt -> false
                          | _ -> true)
            | None -> false)

module Float =
 struct
  (** val transform_quiet_pl : nan_pl -> nan_pl **)

  let transform_quiet_pl pl =
    Pos.coq_lor pl
      (iter_nat (fun x -> Coq_xO x) (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S
        (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S (S
        O))))))))))))))))))))))))))))))))))))))))))))))))))) Coq_xH)

  (** val neg_pl : bool -> nan_pl -> bool * nan_pl **)

  let neg_pl s pl =
    ((negb s), pl)

  (** val abs_pl : bool -> nan_pl -> bool * nan_pl **)

  let abs_pl _ pl =
    (false, pl)

  (** val binop_pl : binary64 -> binary64 -> bool * nan_pl **)

  let binop_pl x y =
    match x with
    | B754_nan (s1, pl1) ->
      (match y with
       | B754_nan (s2, pl2) ->
         if choose_binop_pl_64 s1 pl1 s2 pl2
         then (s2, (transform_quiet_pl pl2))
         else (s1, (transform_quiet_pl pl1))
       | _ -> (s1, (transform_quiet_pl pl1)))
    | _ -> (match y with
            | B754_nan (s2, pl2) -> (s2, (transform_quiet_pl pl2))
            | _ -> default_pl_64)

  (** val zero : float **)

  let zero =
    B754_zero false

  (** val neg : float -> float **)

  let neg =
    coq_Bopp (Zpos (Coq_xI (Coq_xO (Coq_xI (Coq_xO (Coq_xI Coq_xH)))))) (Zpos (Coq_xO (Coq_xO (Coq_xO (Coq_xO
      (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO Coq_xH))))))))))) neg_pl

  (** val abs : float -> float **)

  let abs =
    coq_Babs (Zpos (Coq_xI (Coq_xO (Coq_xI (Coq_xO (Coq_xI Coq_xH)))))) (Zpos (Coq_xO (Coq_xO (Coq_xO (Coq_xO
      (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO Coq_xH))))))))))) abs_pl

  (** val add : float -> float -> float **)

  let add =
    coq_Bplus (Zpos (Coq_xI (Coq_xO (Coq_xI (Coq_xO (Coq_xI Coq_xH)))))) (Zpos (Coq_xO (Coq_xO (Coq_xO (Coq_xO
      (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO Coq_xH))))))))))) binop_pl Coq_mode_NE

  (** val sub : float -> float -> float **)

  let sub =
    coq_Bminus (Zpos (Coq_xI (Coq_xO (Coq_xI (Coq_xO (Coq_xI Coq_xH)))))) (Zpos (Coq_xO (Coq_xO (Coq_xO (Coq_xO
      (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO Coq_xH))))))))))) binop_pl Coq_mode_NE

  (** val mul : float -> float -> float **)

  let mul =
    coq_Bmult (Zpos (Coq_xI (Coq_xO (Coq_xI (Coq_xO (Coq_xI Coq_xH)))))) (Zpos (Coq_xO (Coq_xO (Coq_xO (Coq_xO
      (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO Coq_xH))))))))))) binop_pl Coq_mode_NE

  (** val div : float -> float -> float **)

  let div =
    coq_Bdiv (Zpos (Coq_xI (Coq_xO (Coq_xI (Coq_xO (Coq_xI Coq_xH)))))) (Zpos (Coq_xO (Coq_xO (Coq_xO (Coq_xO
      (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO Coq_xH))))))))))) binop_pl Coq_mode_NE

  (** val compare : float -> float -> Datatypes.comparison option **)

  let compare f1 f2 =
    coq_Bcompare (Zpos (Coq_xI (Coq_xO (Coq_xI (Coq_xO (Coq_xI Coq_xH)))))) (Zpos (Coq_xO (Coq_xO (Coq_xO
      (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO Coq_xH))))))))))) f1 f2

  (** val cmp : comparison -> float -> float -> bool **)

  let cmp c f1 f2 =
    cmp_of_comparison c (compare f1 f2)

  (** val to_long : float -> Int64.int option **)

  let to_long f =
    option_map Int64.repr
      (coq_ZofB_range (Zpos (Coq_xI (Coq_xO (Coq_xI (Coq_xO (Coq_xI Coq_xH)))))) (Zpos (Coq_xO (Coq_xO (Coq_xO
        (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO Coq_xH))))))))))) f Int64.min_signed
        Int64.max_signed)

  (** val from_parsed : positive -> positive -> coq_Z -> float **)

  let from_parsed base intPart expPart =
    coq_Bparse (Zpos (Coq_xI (Coq_xO (Coq_xI (Coq_xO (Coq_xI Coq_xH)))))) (Zpos (Coq_xO (Coq_xO (Coq_xO (Coq_xO
      (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO (Coq_xO Coq_xH))))))))))) base intPart expPart

  (** val to_bits : float -> Int64.int **)

  let to_bits f =
    Int64.repr (bits_of_b64 f)

  (** val of_bits : Int64.int -> float **)

  let of_bits b =
    b64_of_bits (Int64.unsigned b)
 end
